sklearn/utils/validation.py

```diff
@@ -1361,6 +1361,59 @@ def _allclose_dense_sparse(x, y, rtol=1e-7, atol=1e-9):
                      "matrix and an array")
 
 
+def _check_precomputed_gram_matrix(X, precompute, X_offset, X_scale,
+                                   rtol=1e-7,
+                                   atol=1e-5):
+    """Computes a single element of the gram matrix and compares it to
+    the corresponding element of the user supplied gram matrix.
+
+    If the values do not match a ValueError will be thrown.
+
+    Parameters
+    ----------
+    X : array-like of shape (n_samples, n_features)
```

Comment:
It seems that we expect X to have `shape` so it should be a `ndarray`. Would it make sense to support sparse matrices?

Reply:
After checking the code under, I think that we want to support only `ndarray`. So you can switch from `array-like` to `ndarray`.

Reply:
here and below